saleor/webhook/observability_reporter/__init__.py

```diff
@@ -0,0 +1,139 @@
+import logging
+from contextlib import contextmanager
+from contextvars import ContextVar
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Dict, Generator, List, Optional
+
+import opentracing
+import opentracing.tags
+from django.conf import settings
+from graphql import GraphQLDocument
+
+from ...webhook.event_types import WebhookEventAsyncType
+from .buffer import put_event
+from .exceptions import ObservabilityError
+from .payloads import generate_api_call_payload, generate_event_delivery_attempt_payload
+from .utils import webhooks_for_event_exists
+
+if TYPE_CHECKING:
+    from datetime import datetime
+
+    from django.http import HttpRequest, HttpResponse
+
+    from ...core.models import EventDeliveryAttempt
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class GraphQLOperationResponse:
+    name: Optional[str] = None
+    query: Optional[GraphQLDocument] = None
+    variables: Optional[Dict] = None
+    result: Optional[Dict] = None
+    status_code: int = 200
+
+
+class ApiCallResponse:
+    def __init__(self, request: "HttpRequest"):
+        self.request = request
+        self.gql_operations: List[GraphQLOperationResponse] = []
+        self.response: Optional["HttpResponse"] = None
+        self._reported = False
+
+    def report(self) -> bool:
+        if self._reported or not settings.OBSERVABILITY_ACTIVE:
+            return False
+        self._reported = True
+        if (
+            not getattr(self.request, "app", None)
+            and not settings.OBSERVABILITY_REPORT_ALL_API_CALLS
+        ):
+            return False
+        tracer = opentracing.global_tracer()
+        with tracer.start_active_span("observability_report_api_call") as scope:
+            scope.span.set_tag(opentracing.tags.COMPONENT, "observability")
+            if self.response is None:
+                logger.info("[Observability] HttpResponse not provided")
+                return False
+            event_type = WebhookEventAsyncType.OBSERVABILITY
+            if webhooks_for_event_exists(event_type):
```

Comment:
Wouldn't be good to create a cache or something that could reduce the amount of DB queries? Like, this function will add 2-3 queries to each request.

Reply:
That's the last thing on my list of reporter optimizations, but I decided to skip it. It should add one DB query per request coming from apps only ([I tried to confirm it with this test](https://github.com/saleor/saleor/pull/9104/files#diff-67ea1df1ee3831a253d7ad11e34162acecd7f06af48b31a1ff00f2a9b397d440R38-R50)). But yes, that is still a lot of additional DB queries, which we can be reduced by adding caching. If you recommend it, I will do that.

Reply:
Caching is not free. If we want to replace an SQL query with a cache query, make sure it's faster.

Reply:
I will try to measure it, and we will decide